### ISSUE TITLE
Next.js Integration

### ISSUE THREAD

**Issue by paularah** [wellyshen/use-places-autocomplete#1036]
I would appreciate a definitive guide on using this with next.js.  The Next.js script tag with a beforeInteractive strategy gives the error ` use-places-autocomplete: Google Maps Places API library must be loaded.` The native script tag only works with asynchronous loading and occasionally freezes the input field.

**Comment by clearly-outsane** [comment]
I have this problem too

**Comment by DopamineDriven** [comment]
Per the recent Nextjs shift in best practices for loading beforeInteractive scripts, you can instantiate the library as pictured and it works perfectly. I’m using next@latest, edge middleware, etc etc; works with next@12.x.x
![AA66211F-B6D6-49DC-85FB-90708DD14C6A](https://user-images.githubusercontent.com/46355797/178403676-478344fc-233c-4f7f-85dd-b7423eb753c6.png)

if anyone has tips for using the returned latlng values with Algolia InstantSearch Hooks it’d be much appreciated. I currently have it all working in isolation; but when using both Google places and algolia in the same form the fetching strategies clash I think (fiddling with Promise.all([]) but might say fuck it and use global context, keep the worlds separate). 

That said, it would be dope to see an algolia-instantsearch-hooks use places autocomplete integration 
![4ED8D13C-9C3E-4A1F-88C6-5611F3A8B35A](https://user-images.githubusercontent.com/46355797/178404389-de3b47fd-873e-4d18-b750-be36deace7e1.jpeg)

**Comment by scottschindler** [comment]
@DopamineDriven Thanks. 

Is that code in "_document.js"? I have similar code in document and it's not working for me. 

```
import { Html, Head, Main, NextScript } from "next/document";
import Script from "next/script";

export default function Document() {
  return (
    <Html>
      <Head />
      <body>
        <Main />
        <NextScript />
        <Script
          id="googlemaps"
          type="text/javascript"
          strategy="beforeInteractive"
          src="https://maps.googleapis.com/maps/api/js?key=KEY-4Q&libraries=places"
        />
      </body>
    </Html>
  );
}
```

**Comment by dgaitsgo** [comment]
> @DopamineDriven Thanks.
> 
> Is that code in "pages/_document.js"? I have similar code in document and it's not working for me.
> 
> ```
> import { Html, Head, Main, NextScript } from "next/document";
> import Script from "next/script";
> 
> export default function Document() {
>   return (
>     <Html>
>       <Head />
>       <body>
>         <Main />
>         <NextScript />
>         <Script
>           id="googlemaps"
>           type="text/javascript"
>           strategy="beforeInteractive"
>           src="https://maps.googleapis.com/maps/api/js?key=KEY-4Q&libraries=places"
>         />
>       </body>
>     </Html>
>   );
> }
> ```

I didn't want the script available everywhere so only this helped resolve the issue: https://github.com/wellyshen/use-places-autocomplete/discussions/883#discussioncomment-2791354

**Comment by avaneeshtripathi** [comment]
**Reason:**
This issue occurs when `usePlacesAutocomplete` is initialized before the Google Maps script is loaded.

**Solution:**
This can be accomplished with any of the strategies.
1. `beforeInteractive` will anyways work as the google map script is loaded initially.
2. if we have issues with `afterInteractive` or `lazyOnLoad`. we can do the following:

A. We set a flag somewhere in redux/zustand/context when the google maps script is loaded. The placement of script tag will depend on the use cases like:
- If we are using external stores like `zustand` we can place the script in `_document.tsx`
- In case of `context`, the script needs to be places within the boundaries of context to be able to set the flag there.
```
<Script
      id="googlemaps"
      type="text/javascript"
      strategy="lazyOnLoad"
      src="https://maps.googleapis.com/maps/api/js?key=KEY-4Q&libraries=places"
      onLoad={() => useMapStore.setState({ isMapScriptLoaded: true })}
/>
```

B. Wherever we are using `usePlacesAutoComplete`, we [delay initialization of the hook](https://github.com/wellyshen/use-places-autocomplete#lazily-initializing-the-hook) based on our flag:
```
const isMapScriptLoaded = useMapStore((state) => state.isMapScriptLoaded);

const { init } = usePlacesAutocomplete({ initOnMount: false });

useEffect(() => {
    if (isMapScriptLoaded) {
        init();
    }
}, [isMapScriptLoaded]);
```

The above example use zustand store, same can be implemented for redux or context as well.

**Comment by waldothedeveloper** [comment]
Hey there: @paularah 
What I have on my NextJS app v. 12.2.5 and seems to be working is putting the <Script /> from NextJs like this: 

![image](https://user-images.githubusercontent.com/16472274/200993631-36dc1250-caed-4dbe-be52-ec61bff2b8f3.png)


However, this doesn't work for me when I put it on a single page inside the `pages` folder. So I guess the downside to it is that it will load mostly on every page, perhaps hurting performance, but I'm not 100% sure, so I don't want to claim things like that. 

In short, putting it on `_document.js` should work!

**Comment by nyfever007** [comment]
I had same problem. 
After some digging, I was able to work around using "use-google-maps-script" package hook.

Using this package, I was able to wait until script is loaded and then call the component to usePlacesAutocomplete.

**Comment by paularah** [comment]
@nyfever007 any chance you can paste a snippet? I tried reproducing your approach with no success.

**Comment by nyfever007** [comment]
> @nyfever007 any chance you can paste a snippet? I tried reproducing your approach with no success.

`import { Libraries, useGoogleMapsScript } from 'use-google-maps-script';
const libraries: Libraries = ['places'];

const App = () => {
  const { isLoaded, loadError } = useGoogleMapsScript({
    googleMapsApiKey: process.env.NEXT_PUBLIC_GOOLGE_MAPS_API_KEY!,
    libraries,
  });

return(
 <div>
  {isLoaded && (
          <AutoPlacesAddress setSelectedAddress={setSelectedAddress}    defaultValue={null}  />
        )}
</div>
  )
}

export default App
`

`import { useRef } from 'react';
import usePlacesAutocomplete, {
  getDetails,
  getGeocode,
  getLatLng,
} from 'use-places-autocomplete';
import { useOnClickOutside } from 'usehooks-ts';

type Props = {
  setSelectedAddress: React.Dispatch<React.SetStateAction<AddressFormValues>>;
  defaultValue: string | null;
};
const AutoPlacesAddress = ({ setSelectedAddress, defaultValue }: Props) => {


  const {
    ready,
    value,
    suggestions: { status, data },
    setValue,
    clearSuggestions,
  } = usePlacesAutocomplete({
    requestOptions: {
      /* Define search scope here */
    },
    debounce: 500,
  });

  const handleInput = (e: React.ChangeEvent<HTMLInputElement>): void => {

    setValue(e.target.value);
  };

  const ref = useRef(null);
  const handleClickOutside = () => {
 // use useOnClickOutside hook to close the suggestions when user clicks outside of the component
    clearSuggestions();
  };

  useOnClickOutside(ref, handleClickOutside);

 
  const handleSelect =
    (suggestion: { description: string; place_id: string }) => async () => {
      // When user selects a place, we can replace the keyword without request data from API
      // by setting the second parameter to "false"
      setValue(suggestion.description, false);
      const placeId = {
        placeId: suggestion.place_id,
      };

      clearSuggestions();

      try {
     
        const latlng = await getGeocode({ address: suggestion.description });
        const { lat, lng } = getLatLng(latlng[0]);
        const detail: any = await getDetails(placeId);
  
      
        setSelectedAddress({
          address: suggestion.description,
          lat: lat,
          lng: lng,
        });
      } catch (error) {
        console.log(error);
      }

  
    };

  const renderSuggestions = () =>
    data.map((suggestion) => {
      const {
        place_id,
        structured_formatting: { main_text, secondary_text },
      } = suggestion;

      return (
        <li
          key={place_id}
          onClick={handleSelect(suggestion)}
          className='w-full cursor-pointer border-b py-1 px-4  last-of-type:border-0 hover:bg-indigo-500 hover:text-white'
        >
          <div>
            {main_text} <small>{secondary_text}</small>
          </div>
        </li>
      );
    });

  //   if (!isLoaded) return null;
  //   if (loadError) return <div>Error Loading</div>;
  return (
    <>
      <div ref={ref} className='relative'>
        <input
          value={value}
          onChange={handleInput}
          disabled={!ready}
          placeholder={defaultValue!}
          className='form-input w-full rounded-md border-neutral-300 outline-none focus:border-dark-900 focus:outline-0 focus:ring-0'
        />
        {/* We can use the "status" to decide whether we should display the dropdown or not */}
        {status === 'OK' && (
          <ul className='absolute top-10 w-full rounded-md border bg-neutral-50'>
            {renderSuggestions()}
          </ul>
        )}
      </div>
    </>
  );
};

export default AutoPlacesAddress;`


By using use-google-maps-script package you can wait until script is loaded then call your address component.

**Comment by pckilgore** [comment]
If you want to dynamically load in the Google script when component mounts, this worked for me:

```tsx
// hook
  const { init /*, ...rest */ } = usePlacesAutocomplete({
    initOnMount: false,
  });


// component
<Script
  src="https://maps.googleapis.com/maps/api/js?key=${GOOGLE_PLACES_KEY}&libraries=places"
  onReady={init}
/>
```

**Comment by agboolaidris** [comment]
Adding defer solve this issue for me
<Script
          defer
          id="googlemaps"
          src="https://maps.googleapis.com/maps/api/js?key=''&libraries=places"
          strategy="beforeInteractive"
          type="text/javascript"
        />

**Comment by KarenBoyakhchyan** [comment]
> Per the recent Nextjs shift in best practices for loading beforeInteractive scripts, you can instantiate the library as pictured and it works perfectly. I’m using next@latest, edge middleware, etc etc; works with [next@12.x.x](mailto:next@12.x.x) ![AA66211F-B6D6-49DC-85FB-90708DD14C6A](https://user-images.githubusercontent.com/46355797/178403676-478344fc-233c-4f7f-85dd-b7423eb753c6.png)
> 
> if anyone has tips for using the returned latlng values with Algolia InstantSearch Hooks it’d be much appreciated. I currently have it all working in isolation; but when using both Google places and algolia in the same form the fetching strategies clash I think (fiddling with Promise.all([]) but might say fuck it and use global context, keep the worlds separate).
> 
> That said, it would be dope to see an algolia-instantsearch-hooks use places autocomplete integration ![4ED8D13C-9C3E-4A1F-88C6-5611F3A8B35A](https://user-images.githubusercontent.com/46355797/178404389-de3b47fd-873e-4d18-b750-be36deace7e1.jpeg)

Did you have working like this?

![Screen Shot 2023-05-11 at 20 09 10](https://github.com/wellyshen/use-places-autocomplete/assets/109095129/952cb79a-0174-4a34-aa52-75aa3b958258)
Thanks

**Comment by picozzimichele** [comment]
@avaneeshtripathi thanks a lot! Your solution with init() worked well. Had issue waiting for loads on refresh and this solves it!

**Comment by ThePiyushAggarwal** [comment]
In my case, I had `loading=async` in the script source because of the console warning. I removed it and it is working normally now.

**Comment by bill-zhanxg** [comment]
This is how I do it if anyone have the same question (supports lazy loading across different pages):

https://blog.bill-zhanxg.com/google-map-places-autocomplete-with-nextjs
